<a href="https://colab.research.google.com/github/landinh0804/ai4g-portfolio-Dylan/blob/main/Term%201/Week%204/homework/Week4_Workshop_Student.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# AI for Good — Minor ADSAI
## Week 4: Strings, Text & Files

Computers don't *read* text — they *count* it. This week you turn messy human text into something a program (and, from next week, a machine-learning model) can actually use: **string methods** to clean text, **splitting & counting** to measure it, and **files** to read real data in and write results out.

**Theme of the week:** *whose words get counted?* Text data is people's voices — and the voices missing from the file are invisible to every model trained on it. This is the very first step of NLP, which is exactly where Week 5 starts.

---

## Learning Wave 1 — The String Toolkit

### 🧠 First: what *is* a string, really?

A **string** is just a **row of characters** — exactly like a list, but every slot holds one character. So everything you already know about lists still works:

- `text[0]` is the first character, `text[-1]` the last
- `len(text)` is how many characters there are
- `"bike" in text` checks whether that word appears

What strings add is a **toolbox of methods** for cleaning and reshaping text. The ones you will use constantly:

| Method | What it does |
|---|---|
| `.strip()` | remove spaces / newlines at the start and end |
| `.lower()` / `.upper()` | change case, so `"Bike"` and `"bike"` count as the same word |
| `.replace(old, new)` | swap one piece of text for another |
| `.split()` | cut a sentence into a **list of words** |
| `" ".join(words)` | glue a list of words back into one sentence |
| `.count("x")` | how many times `"x"` appears |

**The big idea for the whole week:** a computer cannot understand a sentence. The first move in *every* text and AI pipeline is to **clean** the text and **split it into pieces you can count.**

### 🤔 Predict before you run (2 min — pair up!)
What will `repr(raw)` show that `print(raw)` hides? Guess, then run the cell below.

### Exercise 1: Clean one piece of feedback
⏱ **15 minutes** | Slide 7

**Context:** A city collects open-text feedback about a neighbourhood. Raw text is messy — your first job is to clean a single response.

**Steps:**
1. Create `raw = "  The NEW bike lanes are GREAT!! but the lights are too short.  "`.
2. Print it with `repr()` so you can see the spaces, then `strip()` it into `clean`.
3. Make a `lower()`-case version.
4. How many **characters** does `clean` have? How many **words**? (`len()` and `.split()`)
5. Does the feedback mention `"bike"`? And `"lights"`? Use the `in` operator on the lowercase version.
6. `replace()` the `"!!"` with `"."` and print the final, clean sentence.

In [1]:
raw = "  The NEW bike lanes are GREAT!!  "
print('print:', raw)
print('repr:', repr(raw))

print:   The NEW bike lanes are GREAT!!  
repr: '  The NEW bike lanes are GREAT!!  '


In [2]:
# Exercise 1: Clean one piece of feedback
raw = "  The NEW bike lanes are GREAT!! but the lights are too short.  "

# 1-2. Print raw with repr(), then strip() into `clean`
print(repr(raw))
clean = raw.strip()
print(repr(clean))

# 3. Make a lowercase version
lower = clean.lower()
print(lower)

# 4. Count characters and words
print("Characters:", len(clean))
print("Words:", len(clean.split()))

# 5. Check for "bike" and "lights" with `in`
print("Mentions bike:", "bike" in lower)
print("Mentions lights:", "lights" in lower)

# 6. Replace "!!" with "." and print the final sentence
final = clean.replace("!!", ".")
print(final)

'  The NEW bike lanes are GREAT!! but the lights are too short.  '
'The NEW bike lanes are GREAT!! but the lights are too short.'
the new bike lanes are great!! but the lights are too short.
Characters: 60
Words: 12
Mentions bike: True
Mentions lights: True
The NEW bike lanes are GREAT. but the lights are too short.


---
## Learning Wave 2 — From Text to Numbers

One response is easy. But a city gets *thousands*. To find what people care about, you count words across **all** of them — this is called a **bag of words**, and it is the foundation of almost all classic NLP.

The recipe reuses exactly what you already know:
1. `for` over every response (Week 2)
2. `.lower().split()` each response into words (Week 4)
3. the **counting pattern** from Week 3: `counts[word] = counts.get(word, 0) + 1`

### Exercise 2: Which words come up most?
⏱ **15 minutes** | Slide 10

**Given:**
```python
responses = [
    "The bike lanes are great but too narrow",
    "More bike parking please the parking is always full",
    "Traffic lights are too short for pedestrians",
    "Great new park but we need more lights at night",
    "The park is lovely and the bike lanes are a great idea",
]
```

**Steps:**
1. Build a dictionary `counts` of how often each (lowercased) word appears, using the counting pattern.
2. Print every word with its count using `.items()`.
3. Find the **most mentioned** word with the comparison-variable pattern (Weeks 2–3 — no need for `max()` on a dict).
4. **Bonus:** make a `stopwords` list (`"the"`, `"are"`, `"but"`, `"and"`, `"a"`, `"is"`, `"for"`, …) and skip those words with `continue`, so only meaningful words remain.

In [3]:
# Exercise 2: Which words come up most?
responses = [
    "The bike lanes are great but too narrow",
    "More bike parking please the parking is always full",
    "Traffic lights are too short for pedestrians",
    "Great new park but we need more lights at night",
    "The park is lovely and the bike lanes are a great idea",
]

# 1. Count every word (counting pattern). Loop over responses, then over words.
counts = {}
for response in responses:
    for word in response.lower().split():
        counts[word] = counts.get(word, 0) + 1

# 2. Print each word and its count with .items()
for word, n in counts.items():
    print(word, n)

# 3. Find the most mentioned word (comparison-variable pattern)
best_word = ""
best_count = 0
for word, n in counts.items():
    if n > best_count:
        best_word = word
        best_count = n
print("Most mentioned (with stopwords):", best_word, best_count)

# 4. Bonus: skip stopwords with `continue`
stopwords = ["the", "are", "but", "and", "a", "is", "for", "too", "we", "at", "more", "please", "always", "need"]
clean_counts = {}
for response in responses:
    for word in response.lower().split():
        if word in stopwords:
            continue
        clean_counts[word] = clean_counts.get(word, 0) + 1

best_word = ""
best_count = 0
for word, n in clean_counts.items():
    if n > best_count:
        best_word = word
        best_count = n
print("Most mentioned (no stopwords):", best_word, best_count)
print(clean_counts)

the 4
bike 3
lanes 2
are 3
great 3
but 2
too 2
narrow 1
more 2
parking 2
please 1
is 2
always 1
full 1
traffic 1
lights 2
short 1
for 1
pedestrians 1
new 1
park 2
we 1
need 1
at 1
night 1
lovely 1
and 1
a 1
idea 1
Most mentioned (with stopwords): the 4
Most mentioned (no stopwords): bike 3
{'bike': 3, 'lanes': 2, 'great': 3, 'narrow': 1, 'parking': 2, 'full': 1, 'traffic': 1, 'lights': 2, 'short': 1, 'pedestrians': 1, 'new': 1, 'park': 2, 'night': 1, 'lovely': 1, 'idea': 1}


---
## Learning Wave 3 — Reading & Writing Files

So far the data lived *inside* your code. Real data lives in **files**. Python opens files with `open()`, and the safe way is the `with` block — it closes the file for you automatically:

```python
# Writing
with open("notes.txt", "w", encoding="utf-8") as f:
    f.write("first line\n")      # \n = a new line

# Reading, line by line
with open("notes.txt", "r", encoding="utf-8") as f:
    for line in f:
        print(line.strip())      # strip() removes the trailing newline
```

- `"w"` = write (overwrites the whole file), `"a"` = append, `"r"` = read.
- Always pass `encoding="utf-8"` so accents and symbols don't break.

### Exercise 3: Read feedback from a file
⏱ **20 minutes** (incl. reflection) | Slide 13

**Steps:**
1. Write the lines below to a file `feedback.txt` (use `writelines()`; each line already ends in `\n`).
2. Open `feedback.txt` and loop over it line by line. `strip()` each line; if the line is empty, `continue`.
3. Count the **total** responses, and count how many contain the word `"ALERT"`. Print each urgent line.
4. Write a short report to `report.txt`: total responses and number of urgent ones.
5. **Reflection** (markdown cell below).

```python
lines = [
    "The bike lanes are great\n",
    "ALERT broken glass on the path near the school\n",
    "More lights please it is too dark at night\n",
    "\n",
    "ALERT the crossing feels unsafe for children\n",
    "Lovely new park thank you\n",
]
```

In [4]:
# Exercise 3: Read feedback from a file
# 1. Write the data file
lines = [
    "The bike lanes are great\n",
    "ALERT broken glass on the path near the school\n",
    "More lights please it is too dark at night\n",
    "\n",
    "ALERT the crossing feels unsafe for children\n",
    "Lovely new park thank you\n",
]
with open("feedback.txt", "w", encoding="utf-8") as f:
    f.writelines(lines)

# 2 & 3. Read it back, skip empty lines, count total + urgent, print urgent lines
total = 0
urgent = 0
with open("feedback.txt", "r", encoding="utf-8") as f:
    for line in f:
        line = line.strip()
        if line == "":
            continue
        total += 1
        if "ALERT" in line:
            urgent += 1
            print("URGENT:", line)

print("Total responses:", total)
print("Urgent responses:", urgent)

# 4. Write a small report to report.txt
with open("report.txt", "w", encoding="utf-8") as f:
    f.write("Total responses: " + str(total) + "\n")
    f.write("Urgent responses: " + str(urgent) + "\n")

URGENT: ALERT broken glass on the path near the school
URGENT: ALERT the crossing feels unsafe for children
Total responses: 5
Urgent responses: 2


#### ✍️ Reflection
A text file feels neutral — it is "just the data". But **who is in this file, and who is not?**

Write 2–3 sentences: **Whose feedback ends up in a file like this, and whose never does? If a city (or an AI model) only learns from the text it receives, what happens to the people who never wrote anything down?**

> Mostly people who have time, digital access, confidence in the local language, and trust that someone will listen end up in the file. People who are elderly, have no internet, speak another language, are busy with several jobs, or have given up on being heard never appear. If a city or AI model only learns from the text it receives, those people become invisible, so decisions and models get tuned to the loudest and best-connected groups while everyone else's needs are treated as if they don't exist.

---
## Big Individual Assignment — 135 minutes

### 📊 Community Feedback Analyzer

**Goal:** Build a program that **reads a file of citizen feedback, analyses the text, and writes a report** — a tool a city or NGO could actually use to *listen at scale*.

> **This assignment is more open than before — there is no step-by-step recipe.** You decide how to structure your program (functions, a menu, or one pipeline). What matters is that your solution **uses what you learned today**, end to end.

#### ✅ Your program **must** use all of these (this week's skills):
- **Read** the feedback from a file with a `with open(...)` block
- **Clean** the text with string methods (`strip()`, `lower()`, `replace()` and/or `.strip(".,!?")`)
- **Split** responses into words with `.split()`
- **Count** word frequencies with a dictionary and the counting pattern
- **Write** a readable report to a new file
- (Reuse freely from Weeks 1–3: `if/else`, loops, functions, lists, dicts.)

#### 📋 What the report must contain (the *what* — the *how* is yours):
1. How many responses were analysed, and the **average number of words** per response.
2. The **top 5 keywords** (most frequent words, ignoring stopwords).
3. The number of **urgent** responses (containing words like `unsafe`, `broken`, `alert`, `danger`) — and list them.
4. Save the whole report to `feedback_report.txt` **and** print it to the screen.

#### 🌟 Bonus (for fast finishers):
- A tiny **sentiment score**: count words from a `positive` list vs. a `negative` list and report the mood.
- Let the user **search** for a word and report how many responses mention it.
- Use `import re` and `re.findall(r"\d+", text)` to pull out any **numbers** people mentioned.

#### 💡 Tips:
1. Build it in pieces: first just read + print the file, then add counting, then the report. Test after each piece.
2. A `KeyError` means you reached for a key that isn't there — reach for `.get()` instead.
3. Write one small helper function per job (`read_responses`, `count_words`, …). It makes the program readable and is exactly how real text pipelines are built.

You will need a feedback file to analyse. You can reuse `feedback.txt` from Exercise 3, or create your own with more lines.

#### ✍️ Reflection (markdown cell below).

In [5]:
# Big Assignment: Community Feedback Analyzer
import re

STOPWORDS = ["the", "are", "but", "and", "a", "an", "is", "for", "too", "we", "at", "it", "to",
             "of", "in", "on", "with", "more", "please", "this", "that", "there", "they", "be",
             "have", "has", "was", "were", "i", "you", "our", "my", "very", "so", "not", "by"]
URGENT_WORDS = ["unsafe", "broken", "alert", "danger", "dangerous", "urgent", "emergency"]
POSITIVE = ["great", "good", "lovely", "nice", "thank", "thanks", "love", "clean", "safe", "happy"]
NEGATIVE = ["broken", "unsafe", "dark", "dirty", "narrow", "danger", "bad", "full", "short", "alert"]


def read_responses(filename):
    """Read a file and return a list of non-empty, stripped lines."""
    responses = []
    with open(filename, "r", encoding="utf-8") as f:
        for line in f:
            line = line.strip()
            if line == "":
                continue
            responses.append(line)
    return responses


def clean_words(text):
    """Lowercase, strip punctuation, and split a response into words."""
    words = []
    for word in text.lower().split():
        word = word.strip(".,!?;:\"'()")
        if word != "":
            words.append(word)
    return words


def count_words(responses):
    """Count keyword frequencies, ignoring stopwords."""
    counts = {}
    for response in responses:
        for word in clean_words(response):
            if word in STOPWORDS:
                continue
            counts[word] = counts.get(word, 0) + 1
    return counts


def top_keywords(counts, n=5):
    """Return the n most frequent (word, count) pairs."""
    items = list(counts.items())
    items.sort(key=lambda pair: (-pair[1], pair[0]))
    return items[:n]


def average_words(responses):
    if len(responses) == 0:
        return 0
    total = 0
    for response in responses:
        total += len(clean_words(response))
    return total / len(responses)


def find_urgent(responses):
    urgent = []
    for response in responses:
        words = clean_words(response)
        for w in URGENT_WORDS:
            if w in words:
                urgent.append(response)
                break
    return urgent


def sentiment(responses):
    pos = 0
    neg = 0
    for response in responses:
        for word in clean_words(response):
            if word in POSITIVE:
                pos += 1
            if word in NEGATIVE:
                neg += 1
    if pos > neg:
        mood = "mostly positive"
    elif neg > pos:
        mood = "mostly negative"
    else:
        mood = "balanced"
    return pos, neg, mood


def mentions(responses, word):
    word = word.lower()
    n = 0
    for response in responses:
        if word in clean_words(response):
            n += 1
    return n


def numbers_mentioned(responses):
    found = []
    for response in responses:
        found.extend(re.findall(r"\d+", response))
    return found


def build_report(responses):
    counts = count_words(responses)
    urgent = find_urgent(responses)
    pos, neg, mood = sentiment(responses)
    lines = []
    lines.append("COMMUNITY FEEDBACK REPORT")
    lines.append("=" * 30)
    lines.append("Responses analysed: " + str(len(responses)))
    lines.append("Average words per response: " + str(round(average_words(responses), 1)))
    lines.append("")
    lines.append("Top 5 keywords:")
    for word, n in top_keywords(counts, 5):
        lines.append("  " + word + ": " + str(n))
    lines.append("")
    lines.append("Urgent responses: " + str(len(urgent)))
    for u in urgent:
        lines.append("  - " + u)
    lines.append("")
    lines.append("Sentiment: " + str(pos) + " positive words, " + str(neg) + " negative words -> " + mood)
    nums = numbers_mentioned(responses)
    lines.append("Numbers mentioned: " + (", ".join(nums) if nums else "none"))
    return "\n".join(lines)


def main(filename="feedback.txt"):
    responses = read_responses(filename)
    report = build_report(responses)
    print(report)
    with open("feedback_report.txt", "w", encoding="utf-8") as f:
        f.write(report + "\n")
    return responses


responses = main()

# Bonus: search (change the word here, or use input() when running interactively)
query = "park"
print("\nResponses mentioning '" + query + "':", mentions(responses, query))

COMMUNITY FEEDBACK REPORT
Responses analysed: 5
Average words per response: 7.0

Top 5 keywords:
  alert: 2
  bike: 1
  broken: 1
  children: 1
  crossing: 1

Urgent responses: 2
  - ALERT broken glass on the path near the school
  - ALERT the crossing feels unsafe for children

Sentiment: 3 positive words, 5 negative words -> mostly negative
Numbers mentioned: none

Responses mentioning 'park': 1


#### 🌍 Bonus Reflection
**Your analyzer decides what counts as a "keyword" and what counts as "urgent". Those choices shape what the city sees first. Who should decide which words are treated as urgent — and what gets missed when a machine ranks feedback by word frequency?**

> The people affected should help decide, not only the programmer: residents, community groups, and city staff who deal with safety issues should agree on the urgent list, and it should be reviewed regularly. Ranking by word frequency misses rare but serious issues (one report of a broken crossing outweighs ten "nice park" comments), feedback that uses different words or another language, sarcasm and context ("not unsafe" still matches "unsafe"), and anything from people who wrote less or never wrote at all. So a human should always read the urgent and unusual responses instead of trusting the counts.

In [6]:
# Homework: run the analyzer on real public text
# Dataset: "Sentiment Labelled Sentences" (Kotzias et al., 2015), UCI Machine Learning Repository.
# It contains 1000 real Yelp restaurant reviews. Each line is: review text <TAB> label (1 = positive, 0 = negative).
import urllib.request, zipfile, io

URL = "https://archive.ics.uci.edu/static/public/331/sentiment+labelled+sentences.zip"
with urllib.request.urlopen(URL) as r:
    z = zipfile.ZipFile(io.BytesIO(r.read()))
raw_lines = z.read("sentiment labelled sentences/yelp_labelled.txt").decode("utf-8").splitlines()

# Keep only the review text (drop the label after the tab) and save it as a feedback file
with open("yelp_feedback.txt", "w", encoding="utf-8") as f:
    for line in raw_lines:
        text = line.split("\t")[0].strip()
        if text != "":
            f.write(text + "\n")

# Run the exact same analyzer from the big assignment on the real data
yelp = main("yelp_feedback.txt")

# Extra checks for the reflection below
print("\nResponses mentioning 'service':", mentions(yelp, "service"))
print("Responses mentioning 'not':", mentions(yelp, "not"))
labels = [line.split("\t")[-1] for line in raw_lines if "\t" in line]
print("Real labels in the dataset ->", labels.count("1"), "positive,", labels.count("0"), "negative")


COMMUNITY FEEDBACK REPORT
Responses analysed: 1000
Average words per response: 10.9

Top 5 keywords:
  food: 124
  place: 106
  good: 95
  service: 83
  great: 70

Urgent responses: 0

Sentiment: 229 positive words, 29 negative words -> mostly positive
Numbers mentioned: 4, 23, 5, 10, 5, 2, 30, 2, 2, 40, 3, 2, 20, 1979, 4, 10, 30, 7, 8, 100, 2, 1, 12, 20, 1, 4, 00, 17, 5, 30, 5, 10, 90, 10, 2, 40, 10, 1, 5, 3, 4, 3, 3, 20, 35, 70, 40, 45, 8, 100, 15, 3, 40, 30, 1, 35, 5, 5, 1, 2007, 2, 6, 8, 5, 30, 30, 2, 5, 1, 7, 85, 40, 4, 20, 12, 35, 11, 99, 1, 2

Responses mentioning 'service': 83
Responses mentioning 'not': 112
Real labels in the dataset -> 500 positive, 500 negative


#### 🏠 Homework — what does the analyzer reveal on real text, and what does it miss?

**Data:** 1000 real Yelp restaurant reviews from the UCI *Sentiment Labelled Sentences* dataset (Kotzias et al., 2015). I ran the exact same analyzer from the big assignment on it, without changing anything.

**What it reveals:** the top keywords (*food, place, good, service, great*) show straight away what people talk about most: the food first, then the place and the service. "Service" comes up in 83 of the 1000 reviews, so a restaurant owner would learn quickly that service matters almost as much as the food. That works well because counting words is good at showing **what topics** people mention.

**What it misses:** my sentiment score said the reviews are *mostly positive* (229 positive words vs 29 negative), but the dataset is actually **exactly 500 positive and 500 negative**. The analyzer badly underestimates the negative reviews, for three reasons:
1. **Negation:** "not good" counts as a *positive* word. 112 reviews contain "not", and my code can't see that it flips the meaning.
2. **Missing vocabulary:** my negative list was written for city feedback (*broken, unsafe, dark…*), not for restaurants, so words like *bland, rude, slow, cold* or *overpriced* are invisible.
3. **Urgency doesn't transfer:** 0 urgent responses, because none of my urgent words fit restaurant reviews, even if someone wrote about getting sick.

**Lesson:** word counting tells you *what* people talk about, but not reliably *how they feel*. The word lists only know what the programmer put in them, which is the same problem as this week's theme: whatever isn't in the list (or the file) is invisible. Next week's machine learning tries to fix this by *learning* which words matter from labelled examples, instead of me guessing them.

---
## 5 Things to Remember

1. **A string is a row of characters** — indexing, slicing, `len()` and `in` work just like with lists.
2. **Clean before you count** — `.strip()`, `.lower()` and `.replace()` turn messy text into something comparable (`"Bike!"` and `"bike"` become the same).
3. **`.split()` turns a sentence into a list of words; `" ".join()` turns it back** — this is *tokenisation*, the first step of every NLP system.
4. **Counting words = the Week 3 counting pattern** — `counts[word] = counts.get(word, 0) + 1` is a "bag of words".
5. **Files in, results out** — `with open(...) as f:` reads real data and writes reports safely. And remember: *the words missing from the file are missing from everything you build on it.*

---
**Next week:** Machine Learning basics — KNN, regression and tree-based models with scikit-learn. The cleaned, counted text from this week is exactly the kind of input a model learns from.  
**Homework:** Finish and polish your Community Feedback Analyzer · Find one short piece of real public text (reviews, comments, an open dataset) and run your analyzer on it — what does it reveal, and what does it miss?  
**Friday:** Hackathon 4 — you find out the tool and the SDG on the day.